core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1508,7 +1507,7 @@ private void createStageExecution(
                         memoryEstimatorFactory.createPartitionMemoryEstimator(session, fragment, planFragmentLookup),
                         outputStatsEstimator,
                         // do not retry coordinator only tasks
-                        coordinatorStage ? 1 : maxTaskExecutionAttempts,
+                        shouldRetry(stage) ? 1 : maxTaskExecutionAttempts,
```

Comment:
maxTaskExecutionAttempts bothers me a little. If this is a leaf stage the count would reflect the number of retries for that stage only, but if this is a speculative execution stage, then the number of retries could be much larger (number of dependent tasks + 1) * maxTaskExecutionAttempts. For the coordinator this would be number of dependent tasks * maxTaskExecutionAttempts. Alternatively, we could cancel speculative execution for the subsequent stage if a dependent task fails and wait for all tasks to succeed before starting the next task.

Reply:
Actually it will not be a problem as we are not restarting speculative tasks. If we get a failure on a task which is from stage marked as speculative we will restart it only after all the source stages complete.

https://github.com/trinodb/trino/blob/619d7ef9d600d5ad6669c91905021d67b0d8bd46/core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java#L2476-L2479